File: src/client/java/net/greenjab/fixedminecraft/hud/HUDOverlayHandler.java
package net.greenjab.fixedminecraft.hud;

import net.greenjab.fixedminecraft.FixedMinecraft;
import net.greenjab.fixedminecraft.util.IntPoint;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.renderer.RenderPipelines;
import net.minecraft.util.ARGB;
import net.minecraft.util.Mth;
import net.minecraft.world.effect.MobEffects;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.food.FoodData;

import java.util.Random;
import java.util.Vector;

/** Credit: Squeek502 */
public class HUDOverlayHandler {

    private static final int FOOD_BAR_HEIGHT = 39;

    private static final Vector<IntPoint> foodBarOffsets = new Vector<>();

    private static final Random random = new Random();

    public static void onRender(GuiGraphicsExtractor context) {
        Minecraft mc = Minecraft.getInstance();
        Player player = mc.player;
        assert player != null;
        FoodData stats = player.getFoodData();

        int top = mc.getWindow().getGuiScaledHeight() - FOOD_BAR_HEIGHT;
        int right = mc.getWindow().getGuiScaledWidth() / 2 + 91;

        generateBarOffsets(top, right, mc.gui.getGuiTicks(), player);

        HUDOverlayEvent.Saturation saturationRenderEvent = new HUDOverlayEvent.Saturation(stats.getSaturationLevel(), right, top, context);

        if (!saturationRenderEvent.isCanceled) {
            drawSaturationOverlay(saturationRenderEvent, mc.gui.getGuiTicks());
        }

    }

    private static void drawSaturationOverlay(GuiGraphicsExtractor context, Float saturationLevel, int right, int top, int ticks) {
        float modifiedSaturation = Mth.clamp(saturationLevel, 0.0f, 20.0f);
        int endSaturationBar = Mth.ceil(modifiedSaturation / 2);
        int iconSize = 9;
        for (int i = 0; i < endSaturationBar; i++) {
            IntPoint offset = foodBarOffsets.get(i);
            if (offset == null) continue;
            int x = right + offset.x;
            int y = top + offset.y;

            float u = 0f;
            float effectiveSaturationOfBar = (modifiedSaturation / 2.0F) - i;
            if (effectiveSaturationOfBar > .75) u = 3f * iconSize;
            else if (effectiveSaturationOfBar > .50) u = 2f * iconSize;
            else if (effectiveSaturationOfBar > .25) u = 1f * iconSize;

            context.blit(RenderPipelines.GUI_TEXTURED, FixedMinecraft.id("textures/gui/sprites/stamina.png"), x, y, u, 0, iconSize, iconSize, 36, 18,  ARGB.white(1F));
            if ((saturationLevel <=6 && saturationLevel>0) || Minecraft.getInstance().player.hasEffect(MobEffects.HUNGER))
                context.blit(RenderPipelines.GUI_TEXTURED, FixedMinecraft.id("textures/gui/sprites/stamina.png"), x, y, u, iconSize, iconSize, iconSize, 36, 18,  ARGB.white((float) (Math.sin(ticks/6f) + 1) / 2f));
        }
        if (saturationLevel <=0) {
            for (int i = 0; i < 10; i++) {
                IntPoint offset = foodBarOffsets.get(i);
                if (offset == null) continue;
                context.blit(RenderPipelines.GUI_TEXTURED, FixedMinecraft.id("textures/gui/sprites/stamina.png"),
                        right + offset.x, top + offset.y,  3f * iconSize, iconSize, iconSize, iconSize, 36, 18,  ARGB.white((float) (Math.sin(ticks/3f) + 1) / 2f));
            }
        }
    }

    private static void drawSaturationOverlay(HUDOverlayEvent.Saturation event, int ticks) {
        drawSaturationOverlay(event.context, event.saturationLevel, event.x, event.y, ticks);
    }

    private static void generateBarOffsets(int top, int right, int ticks, Player player) {
        int preferFoodBars = 10;
        var shouldAnimatedFood = false;
        FoodData hungerManager = player.getFoodData();

        // in vanilla saturation level is zero will show hunger animation
        float saturationLevel = hungerManager.getSaturationLevel();
        int foodLevel = hungerManager.getFoodLevel();
        shouldAnimatedFood = saturationLevel <= 0.0F && ticks % (foodLevel * 3 + 1) == 0;

        // hard code in `InGameHUD`
        random.setSeed((ticks * 312871L));

        if (foodBarOffsets.size() != preferFoodBars)
            foodBarOffsets.setSize(preferFoodBars);

        // right alignment, single row
        for (int i = 0; i < preferFoodBars; i++) {
            int x = right - i * 8 - 9;
            int y = top;

            // apply the animated offset
            if (shouldAnimatedFood)
                y += random.nextInt(3) - 1;

            // reuse the point object to reduce memory usage
            var point = foodBarOffsets.get(i);
            if (point == null) {
                point = new IntPoint();
                foodBarOffsets.set(i, point);
            }

            point.x = x - right;
            point.y = y - top;
        }
    }
}

File: src/client/kotlin/net/greenjab/fixedminecraft/hud/HUDOverlayHandler.kt
package net.greenjab.fixedminecraft.hud

import net.greenjab.fixedminecraft.util.ExhaustionHelper
import net.greenjab.fixedminecraft.util.IntPoint
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.render.RenderLayer
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.Identifier
import net.minecraft.util.math.ColorHelper
import java.util.Random
import java.util.Vector
import kotlin.math.ceil


object HUDOverlayHandler {
    private var foodIconsOffset = 0

    private const val FOOD_BAR_HEIGHT = 39

    private var foodBarOffsets: Vector<IntPoint> = Vector<IntPoint>()

    private var random: Random = Random()

    fun onPreRender(context: DrawContext) {
        foodIconsOffset = FOOD_BAR_HEIGHT
        val mc = MinecraftClient.getInstance()
        val player = mc.player
        requireNotNull(player)
        val right = mc.window.scaledWidth / 2 + 91
        val top = mc.window.scaledHeight - foodIconsOffset
        val exhaustion = ExhaustionHelper.getExhaustion(player);
//player.hungerManager.exhaustion
        val renderEvent = HUDOverlayEvent.Exhaustion(exhaustion, right, top, context)

        if (!renderEvent.isCanceled) {
            drawExhaustionOverlay(renderEvent, mc)
        }
    }

    fun onRender(context: DrawContext) {

        val mc: MinecraftClient = MinecraftClient.getInstance()
        val player: PlayerEntity = mc.player!!
        val stats = player.hungerManager

        val top = mc.window.scaledHeight - foodIconsOffset
        val left = mc.window.scaledWidth / 2 - 91
        val right = mc.window.scaledWidth / 2 + 91

        generateBarOffsets(top, left, right, mc.inGameHud.ticks, player)

        val saturationRenderEvent = HUDOverlayEvent.Saturation(stats.saturationLevel, right, top, context)

        if (!saturationRenderEvent.isCanceled) {
            drawSaturationOverlay(saturationRenderEvent, mc)
        }

    }

    private fun drawSaturationOverlay(context: DrawContext, saturationLevel: Float, mc: MinecraftClient, right: Int, top: Int) {

        val modifiedSaturation = Math.max(0.0f, Math.min(saturationLevel, 20.0f))
        val endSaturationBar = (ceil(modifiedSaturation / 2)).toInt()
        val iconSize = 9

        for (i in 0 until endSaturationBar) {
            // gets the offset that needs to be render of icon
            val offset = foodBarOffsets.get(i)
            if (offset == null)
                continue

            val x = right + offset.x
            val y = top + offset.y

            val v = 0f
            var u = 0f

            val effectiveSaturationOfBar = (modifiedSaturation / 2.0F) - i

            if (effectiveSaturationOfBar > .75)
                u = 3f * iconSize
            else if (effectiveSaturationOfBar > .50)
                u = 2f * iconSize
            else if (effectiveSaturationOfBar > .25)
                u = 1f * iconSize
            context.drawTexture(RenderLayer::getGuiTextured, Identifier.of("fixedminecraft", "textures/icons.png"), x, y, u, v, iconSize, iconSize, 256, 256,  ColorHelper.getWhite(1F))
        }

    }

    private fun drawExhaustionOverlay(context: DrawContext, exhaustion: Float, mc: MinecraftClient, right: Int, top: Int) {
        val maxExhaustion = 1.0f
        // clamp between 0 and 1
        val ratio = (exhaustion / maxExhaustion).coerceIn(0f, 1f)
        val width = (ratio * 81).toInt()
        val height = 9


        context.drawTexture(RenderLayer::getGuiTextured, Identifier.of("fixedminecraft", "textures/icons.png"), right - width, top, 81f - width, 18f, width, height, 128, 16,
            ColorHelper.getWhite(1F))
    }


    private fun drawSaturationOverlay(event: HUDOverlayEvent.Saturation, mc: MinecraftClient) {
        drawSaturationOverlay(event.context, event.saturationLevel, mc, event.x, event.y)
    }


    private fun drawExhaustionOverlay(event: HUDOverlayEvent.Exhaustion, mc: MinecraftClient) {
        drawExhaustionOverlay(event.context, event.exhaustion, mc, event.x, event.y)
    }

    private fun generateBarOffsets(top: Int, left: Int, right: Int, ticks: Int, player: PlayerEntity) {
        val preferFoodBars = 10
        var shouldAnimatedFood = false
        val hungerManager = player.hungerManager

            // in vanilla saturation level is zero will show hunger animation
        val saturationLevel = hungerManager.saturationLevel
        val foodLevel = hungerManager.foodLevel
        shouldAnimatedFood = saturationLevel <= 0.0F && ticks % (foodLevel * 3 + 1) == 0


        // hard code in `InGameHUD`
        random.setSeed((ticks * 312871).toLong())

        if (foodBarOffsets.size != preferFoodBars)
            foodBarOffsets.setSize(preferFoodBars)

        // right alignment, single row
        for (i in 0 until preferFoodBars) {
            val x = right - i * 8 - 9
            var y = top

            // apply the animated offset
            if (shouldAnimatedFood)
                y += random.nextInt(3) - 1

            // reuse the point object to reduce memory usage
            var point = foodBarOffsets.get(i)
            if (point == null) {
                point = IntPoint()
                foodBarOffsets.set(i, point)
            }

            point.x = x - right
            point.y = y - top
        }
    }
}
